From 6ddc5f36e9e4fd4fb971d11711c0c9e411bb8a9b Mon Sep 17 00:00:00 2001
From: Henrik Norlin <henrik.norl@gmail.com>
Date: Thu, 29 Jul 2021 22:27:03 +0200
Subject: [PATCH] External id records with company_id

---
 odoo/models.py | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/odoo/models.py b/odoo/models.py
index fdd94544..a5ecad7f 100644
--- a/odoo/models.py
+++ b/odoo/models.py
@@ -682,7 +682,7 @@ class BaseModel(MetaModel('DummyModel', (object,), {'_register': False})):
             )))
             for r in missing
         )
-        fields = ['module', 'model', 'name', 'res_id']
+        fields = ['module', 'model', 'name', 'res_id', 'company_id']
 
         # disable eventual async callback / support for the extent of
         # the COPY FROM, as these are apparently incompatible
@@ -691,11 +691,12 @@ class BaseModel(MetaModel('DummyModel', (object,), {'_register': False})):
         try:
             cr.copy_from(io.StringIO(
                 u'\n'.join(
-                    u"%s\t%s\t%s\t%d" % (
+                    u"%s\t%s\t%s\t%d\t%d" % (
                         modname,
                         record._name,
                         xids[record.id][1],
                         record.id,
+                        record.company_id.id,
                     )
                     for record in missing
                 )),
-- 
2.17.1

